### PR TITLE
Use sockets for communication if available

### PR DESCRIPTION
**Pull request by CharlotteDunois**
**Pull Request description:**
Uses sockets for communication between pool and threads, if available.

**Why should this Pull Request be merged:**
Boosted performance when using sockets.

**Semantic Versioning Classification:**
- [ ] This PR includes major breaking changes (such as method changes)
- [ ] This PR includes **only** documentational changes (such as docblocks, README, etc.)

### CODE REVIEW

File: src/Worker.php [hunk 1]
```diff
@@ -173,9 +208,24 @@ protected function loop() {
...
+            static::$socket->close();
+            static::$socket = null;
+        }
+        
+        // Inform the pool we are exiting the loop
+        $message = new \CharlotteDunois\Phoebe\Message('internal-exit-loop', array());
+        
+        if($this->bag->socket !== null) {
+            // Create fresh event loop which exits when the stream is done writing
+            $loop = \React\EventLoop\Factory::create();
```
Comment (WyriHaximus) [inline src/Worker.php]: This runs inside worker thread right, not the parent thread?

File: src/Messenger.php [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            $boot = $socket->getWorkerSocket();
+            
+            $this->streams->delete($worker->id);
+            $stream->close();
+            
+            if(\is_resource($sock)) {
+                \socket_close($sock);
+            }
+            
+            if(\is_resource($sock)) {
```
[the remark this answers is not in the document]
Comment (CharlotteDunois) [reply src/Messenger.php]: Copypaste error, should be `$boot` 😄

File: src/Pool.php [hunk 3]
```diff
@@ -207,36 +153,72 @@ function attachTimer() {
...
-            $task['reject'](new \RuntimeException('Shutting down pool'));
-        }
-        
-        $this->tasks->clear();
-        
-        while($worker = \array_shift($this->workers)) {
-            $worker->stop();
-            $worker->join();
-        }
+        return (new \React\Promise\Promise(function (callable $resolve, callable $reject) {
```
Comment (CharlotteDunois) [inline src/Pool.php]: I should clean up `$this->workers` by removing already stopped workers, such as

File: src/Pool.php [hunk 4]
```diff
@@ -358,12 +350,16 @@ function submitTaskTo(int $workerID, \CharlotteDunois\Phoebe\AsyncTask $task) {
...
-        if($pos !== false) {
-            unset($this->workers[$pos]);
+        if($worker->isRunning()) {
+            $message = new \CharlotteDunois\Phoebe\Message('internal-worker-exit', null);
+            $this->sendMessageToWorker($worker, $message);
+        } else {
+            $pos = \array_search($worker, $this->workers, true);
+            if($pos !== false) {
+                unset($this->workers[$pos]);
+            }
```
Comment (CharlotteDunois) [inline src/Pool.php]: Call `$this->messenger->removeWorkerSocket` here and clean up

File: src/Worker.php [hunk 5]
```diff
@@ -162,7 +197,7 @@ protected function loop() {
             return $this->loop();
```
Comment (CharlotteDunois) [inline src/Worker.php]: Add a try-catch for the run method

File: src/Pool.php [hunk 6]
```diff
@@ -128,77 +137,14 @@ function __construct(\React\EventLoop\LoopInterface $loop, array $options, strin
...
-                        } elseif((\time() - $task['dateline']) > $this->tasksTimeout && $task['worker']->getStacked() === 0) {
-                            $this->tasks->delete($id);
-                            $task['reject'](new \RuntimeException('Unknown Termination Reason'));
-                        }
-                    }
-                }
-                
-                $this->tasksLastTime = \time();
-            }
+        $this->timer = $this->loop->addPeriodicTimer(10, function () {
```
Comment (CharlotteDunois) [inline src/Pool.php]: Set timer to 30 seconds and only attach if using sockets

File: src/Pool.php [hunk 7]
```diff
@@ -377,6 +373,90 @@ function stopWorker(\CharlotteDunois\Phoebe\Worker $worker) {
...
+     */
+    protected function internalTimerTick() {
+        foreach($this->busyWorkers as $worker) {
+            if($worker->getStacked() === 0) {
+                $this->busyWorkers->delete($worker);
+                $this->freeWorkers->enqueue($worker);
+            }
+        }
+        
+        if(!$this->lastTime || (\time() - $this->lastTime) > 30) {
```
Comment (CharlotteDunois) [inline src/Pool.php]: Remove check here and move it into timerTick for internal timer ticking when using timers

File: src/Pool.php [hunk 8]
```diff
@@ -377,6 +373,90 @@ function stopWorker(\CharlotteDunois\Phoebe\Worker $worker) {
...
+            }
+        }
+    }
+    
+    /**
+     * Internal Timer tick.
+     * @return void
+     */
+    protected function internalTimerTick() {
+        foreach($this->busyWorkers as $worker) {
```
Comment (CharlotteDunois) [inline src/Pool.php]: Move into timerTick

File: src/ResourceSocket.php [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+                    $code = \socket_last_error();
+                    
+                    if($code !== \SOCKET_EWOULDBLOCK) {
+                        throw new \RuntimeException(\socket_strerror($code));
+                    }
+                    
+                    return;
+                }
+                
+                $exceeded = isset($this->data[$this->softLimit - 1]);
```
Comment (CharlotteDunois) [inline src/ResourceSocket.php]: Parens

File: src/Pool.php [hunk 10]
```diff
@@ -400,6 +482,8 @@ protected function spawnWorker() {
      */
     protected function checkAlive(\CharlotteDunois\Phoebe\Worker $worker) {
         if(!$worker->isRunning()) {
+            $this->emit('debug', 'Picked worker '.$worker->id.' is not running, spawning new one...');
+
```
Comment (CharlotteDunois) [inline src/Pool.php]: Also call `$this->messenger->removeWorkerSocket` here

File: src/Worker.php (hunk 1 above)
Comment (WyriHaximus) [reply src/Worker.php]: Thought so, just checking 👍

File: src/Pool.php [hunk 11]
```diff
@@ -207,36 +153,72 @@ function attachTimer() {
...
+                    
+                    while($worker = \array_shift($workers)) {
+                        $worker->join();
+                    }
+                    
+                    $resolve();
+                }
+            };
+            
+            $timer = $this->loop->addTimer(120, function () use ($reject, &$listener) {
```
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply src/Pool.php]: Sounds good, the only thing that feels odd is the 50ms blocking the loop. Is that main or child thread? Your scale solution sounds good to me 👍

[the remark this answers is not in the document]
Comment (WyriHaximus) [reply src/Pool.php]: Cool, any idea how long the main thread blocks when joining, over or under a milisecond?

[the remark this answers is not in the document]
Comment (CharlotteDunois) [reply src/Pool.php]: I spawned 5 workers:
```
Joining... Done: 15043ns
Joining... Done: 11433ns
Joining... Done: 10229ns
Joining... Done: 9628ns
Joining... Done: 10229ns
```

Is this fast enough? :)

[the remark this answers is not in the document]
Comment (WyriHaximus) [reply src/Pool.php]: Perfect 👍

**Pull request merged**